### ISSUE TITLE
jet response definition

### ISSUE THREAD

**Issue by jpata** [jpata/cmssw#13]
In order to reproduce the TDR jet response plot, we need to make sure the jet response in `ParticleFlowDQM.cc` is implemented in the same way as in the JetMET code.
I started to dig through the [JetMET code](https://github.com/cms-jet/JetMETAnalysis) but so far didn't find the exact defition.

As a start, I changed the jet matching from reco->gen to gen->reco [here](https://github.com/jpata/cmssw/commit/9bbe73758d96acba263255869873b7c1c478304a#diff-9259382196170116b9b95bc9521a6bdeR143):

Likely also the matching deltaR has to be changed to conform to hald the cone radius: https://github.com/cms-jet/JetMETAnalysis/blob/master/JetAnalyzers/python/Defaults_cff.py#L33

@juska @hatakeyamak

**Comment by jpata** [comment]
Changed dR, let's see how this works: https://github.com/jpata/cmssw/commit/815f05899eb623b9d98d8a2c7fa43ea6771754d3

**Comment by jpata** [comment]
via Juska:

In summary, we should use the recoJet-genJet pair with the smallest deltaR, with the usual minimum requirement deltaR<0.2.
I don't know how this works in a (high-)pileup environment where a pileup jet could have the smallest deltaR, but let's start with this 'offcial' definition and see if an additional pT requirement would be necessary in say HL-LHC simulations.

Cheers,

Juska

[*]
We match within deltaR / 2.
So 0.2 for AK4 jets.
The matching code is here
https://github.com/cms-jet/JetMETAnalysis/blob/master/JetUtilities/plugins/MatchRecToGen.cc
It starts with reco jets, loops over each gen jet, and stores the jet pair with the smallest deltaR
You can call the plugin as a EDProducer in your own analysis like here
https://github.com/cms-jet/JetMETAnalysis/blob/master/JetAnalyzers/python/addAlgorithm.py#L482-L485
An example of how to use: The response code uses the 'gen2rec' map (aka ref jet to reco jet) here
https://github.com/cms-jet/JetMETAnalysis/blob/master/JetAnalyzers/src/JetResponseAnalyzer.cc#L243
Also here is where the deltaR condition is called (after matching)
https://github.com/cms-jet/JetMETAnalysis/blob/master/JetAnalyzers/src/JetResponseAnalyzer.cc#L261

**Comment by jpata** [comment]
I chose to use `PFB::match` [defined here](https://github.com/cms-sw/cmssw/blob/master/DQMOffline/PFTau/interface/Matchers.h#L18), which is just a simple function call, rather than the `MatchRecToGen` EDProducer, as they implement the same functionality as far as I can tell. `PFB::match` is already merged into CMSSW and doesn't need to be downloaded as a standalone repository, making the setup of the PF validation code easier (just one `git cms-merge-topic`).
I changed the deltaR criterion to 0.2 and ran a new DQM file, see attached.

Here is a DWM file with the raw histograms:
[DQM_V0001_R000000001__Global__CMSSW_X_Y_Z__RECO.root.zip](https://github.com/jpata/cmssw/files/2798319/DQM_V0001_R000000001__Global__CMSSW_X_Y_Z__RECO.root.zip)

Here is a notebook where I draw the total response plot as a function of pt:
[response_plot.html.zip](https://github.com/jpata/cmssw/files/2798322/response_plot.html.zip)

Here is an example pt-dependent response plot in $abs(genjet.eta) < 0.5$.
![response_eta05](https://user-images.githubusercontent.com/69717/51779389-f7cea900-20bb-11e9-97b5-6b12f35dcacf.png)

**Comment by jpata** [comment]
This is something I made to cross-check the code & samples using FlatQCD with no pileup, |eta_gen|<0.5.

The error bars are set according to `TH1::GetStdDevError`, but just looking at them, this seems to underestimate the true statistical uncertainty. I believe Juska had an update in his script, will take a look once I get a chance.

Yes, once everything is in place, we will have one pt-dependent plot per eta bin as specified here: https://github.com/jpata/cmssw/blob/pfvalidation/Validation/RecoParticleFlow/python/particleFlowDQM_cff.py#L27

**Comment by hatakeyamak** [comment]
Okay. I see. Right, stat uncertainties look underestimated now. It will be good to double-check.
Also, are central values of highest three gen-jet-pt points reasonable? They look very close to 0, or looks almost negative. Is it possible to give tick on the right (and top?) sides as well?

**Comment by jpata** [comment]
Thanks for feedback @hatakeyamak! Good point on the high-pt bins: looking at the raw response plots [here](https://github.com/jpata/cmssw/issues/13#issuecomment-457780748), the last bins are empty so the StdDev is returned as 0. This was with ~5000 events (50 jobs, ~2h each). I was running a bigger sample (~50k) as well, but so far had trouble to finish all the jobs due to disk quota issues, as the total RECO output would be 150GB, need to take a look at this.

**Comment by jpata** [comment]
Seems like this question is solved.
